mne/beamformer/tests/test_dics.py

```diff
@@ -570,6 +575,69 @@ def test_apply_dics_timeseries(_load_forward, idx):
         apply_dics_epochs(epochs, filters_vol)
 
 
+@testing.requires_testing_data
+@pytest.mark.parametrize('return_generator', (True, False))
+def test_apply_dics_tfr(return_generator):
+    """Test lcmv applied to time-frequency objects."""
+    rng = np.random.default_rng(11)
+    n_epochs = 8
+    info = read_info(fname_raw)
+    info = pick_info(info, pick_types(info, meg='grad'))
+    forward = mne.read_forward_solution(fname_fwd)
+    freqs = np.arange(8, 10)
+    times = np.linspace(0, 1, int(round(info['sfreq'])))
+    data = rng.random((n_epochs, len(info.ch_names), freqs.size, times.size))
+    data *= 1e-6
+    epochs_tfr = EpochsTFR(info, data, times=times, freqs=freqs)
+    epochs_tfr.apply_baseline((0, 0.5))
+    csd = csd_tfr(epochs_tfr, tmin=0.5, tmax=1)
+    pick_ori = None  # csd doesn't support vector so can't use for testing
+    filters = make_dics(epochs_tfr.info, forward, csd,
+                        pick_ori=pick_ori, reg=0.05,
+                        real_filter=False)
+    with pytest.raises(RuntimeError,
+                       match='Time-frequency data must be complex'):
+        stcs = apply_dics_tfr_epochs(epochs_tfr, filters)
+
+    # simulate amplitude and phase output
+    epochs_tfr.data = epochs_tfr.data + epochs_tfr.data * 1j
+    stcs = apply_dics_tfr_epochs(epochs_tfr, filters, return_generator)
+
+    # FIXME
+    # because of the round-trip conversion to tstep, this fails
+    # assert_allclose(stcs[0][0].times, times)
+    if return_generator:
+        stcs = [[s for s in these_stcs] for these_stcs in stcs]
+    assert len(stcs) == len(epochs_tfr)
+    assert all([len(s) == freqs.size for s in stcs])
+    data_shape = (forward['nsource'],)
+    if pick_ori == 'vector':
+        data_shape += (3,)
+    data_shape += (times.size,)
+    assert all([s.data.shape == data_shape
+                for these_stcs in stcs for s in these_stcs])
+
+    csd = mne.time_frequency.csd_tfr(epochs_tfr)
+    stc, freqs2 = apply_dics_csd(csd, filters)
+    assert_allclose(freqs, freqs2)
+    tfr_stc_data = np.array([[stc.data for stc in epoch_stcs]
+                             for epoch_stcs in stcs])
+    tfr_csd_data = np.zeros(stc.data.shape)
+
+    # convert to csd
+    for idx, epochs_data in enumerate(tfr_stc_data):
+        power = (epochs_data * np.conj(epochs_data)).real
+        power = power.mean(axis=-1)  # mean over time
+        # Scaling by sampling frequency for compatibility with Matlab
```

Comment:
Why compatibility with Matlab?

Reply:
I acutally misunderstood your intentions here first - again, add comments further up to explain what this test will do

Reply:
This is copied verbatim from the csd code so that the same process and normalizations are followed. That's a good question which I might guess that the answer is that the original MNE implementation was a translation from Matlab.

Reply:
Possibly to more closely reproduce what Fieldtrip does?